modyn/selector/internal/selector_strategies/coreset_strategy.py

```diff
@@ -0,0 +1,102 @@
+            config, maximum_keys_in_memory
+        )
+
+        if isinstance(self.presampling_strategy, EmptyPresamplingStrategy) and isinstance(
+            self.downsampling_strategy, EmptyDownsamplingStrategy
+        ):
+            raise ValueError(
+                "Using a Coreset Method without presampling and downsampling is useless. "
+                "You can use NewDataStrategy instead. "
+                "To specify the presampling method add 'presampling_strategy to the pipeline. "
+                "You can use 'downsampling_strategies' to specify the downsampling method."
+            )
+
+    def inform_data(self, keys: list[int], timestamps: list[int], labels: list[int]) -> None:
+        assert len(keys) == len(timestamps)
+        assert len(timestamps) == len(labels)
+
+        self._persist_samples(keys, timestamps, labels)
+
+    def _on_trigger(self) -> Iterable[list[tuple[int, float]]]:
+        for samples in self._get_data():
+            random.shuffle(samples)
+            yield [(sample, 1.0) for sample in samples]
+
+    def _get_data(self) -> Iterable[list[int]]:
+        """Returns all sample
+
+        Returns:
+            list[str]: Keys of used samples
+        """
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            if self.presampling_strategy.requires_trigger_dataset_size():
+                target_size = self._get_dataset_size()
+            else:
+                target_size = None
+            stmt = self.presampling_strategy.get_presampling_query(
+                self._next_trigger_id,
+                self.tail_triggers,
+                self.training_set_size_limit if self.has_limit else None,
+                target_size,
+            )
+
+            for chunk in database.session.execute(stmt).partitions():
+                if len(chunk) > 0:
+                    yield [res[0] for res in chunk]
+                else:
+                    yield []
+
+    def _reset_state(self) -> None:
+        pass  # As we currently hold everything in database (#116), this currently is a noop.
+
+    def _get_dataset_size(self) -> int:
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            return (
+                database.session.query(SelectorStateMetadata.sample_key)
+                .filter(
+                    SelectorStateMetadata.pipeline_id == self._pipeline_id,
+                    SelectorStateMetadata.seen_in_trigger_id >= self._next_trigger_id - self.tail_triggers
+                    if self.tail_triggers is not None
+                    else True,
+                )
+                .count()
+            )
+
+    def get_downsampling_strategy(self) -> str:
+        return self.downsampling_strategy.get_downsampling_strategy()
+
+    def get_downsampling_params(self) -> dict:
+        return self.downsampling_strategy.get_downsampling_params()
+
```

Comment:
similar comment for these with `get`

Reply:
I tried to reduce the gets as much as possible but here I can't remove them